File: 802.find-eventual-safe-states.py
from typing import List


# @leet start
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        # to deal with cycles as well
        # also deal with 1 -> 2 -> 3, 1->3

        n = len(graph)
        is_safe = [False] * n
        in_path = [False] * n

        def dfs(i: int, in_path: list[bool]) -> bool:

            if not graph[i]:
                is_safe[i] = True
                return True
            if in_path[i]:
                is_safe[i] = False
                return False
            if is_safe[i]:
                return True

            for nb in graph[i]:
                in_path[i] = True
                seems_safe = dfs(nb, in_path)
                in_path[i] = False
                if seems_safe:
                    continue
                else:
                    return False

            is_safe[i] = True
            return True

        for i in range(n):
            dfs(i, in_path)
        return [i for i, b in enumerate(is_safe) if b == True]


# @leet end

g = [[1, 2], [2, 3], [5], [0], [5], [], []]
g = [[1, 2, 3, 4], [1, 2], [3, 4], [0, 4], []]
x = Solution().eventualSafeNodes(g)
print(x)

File: 802.find-eventual-safe-states.java
// @leet start

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        // adapted from NeetCode

        int n = graph.length;
        Map<Integer, Boolean> isSafe = new HashMap<>();
        ArrayList<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (this.dfs(i, isSafe, graph)) {
                ans.add(i);
            }
        }
        return ans;

    }

    boolean dfs(int i, Map<Integer, Boolean> isSafe, int[][] graph) {
        if (isSafe.containsKey(i)) {
            return isSafe.get(i);
        }
        isSafe.put(i, false);

        for (int nb : graph[i]) {
            if (!dfs(nb, isSafe, graph)) {
                return false;
            }

        }
        isSafe.put(i, true);
        return true;

    }
}
// @leet end
